### ISSUE TITLE
Crash when attaching item entity to player

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#10912]
##### Minetest version
```
2072afb72b4b3e9c5dcbcec71d824aeae1b35d19
```

aka 5.4.0-dev

##### Summary
If you attach an item entity to a player, the game errors out.

```
2021-02-04 06:56:57: ERROR[Main]: ServerError: AsyncErr: ServerThread::run Lua: Runtime error from mod '*builtin*' in callback luaentity_Step(): ...zzy/src/minetest/git/bin/../builtin/game/item_entity.lua:199: Collision info missing, this is caused by an out-of-date/buggy mod or game
2021-02-04 06:56:57: ERROR[Main]: stack traceback:
2021-02-04 06:56:57: ERROR[Main]: 	[C]: in function 'assert'
2021-02-04 06:56:57: ERROR[Main]: 	...zzy/src/minetest/git/bin/../builtin/game/item_entity.lua:199: in function <...zzy/src/minetest/git/bin/../builtin/game/item_entity.lua:155>
```

##### Steps to reproduce
You want to attach an item entity to yourself.

1. Start DevTest
2. Give yourself the object attacher from the Chest of Everything
3. Drop any item on the ground
4. Punch air with the object attacher to select yourself as parent
5. Punch item entity with the object attacher to select item entity as child. This will call `set_attach`

This will cause the game to error out in the aforementioned error msg.

**Comment by sfan5** [comment]
Is this supposed to be supported? I don't think this is new in 5.4

**Comment by Wuzzy2** [comment]
You're right, this is not new in 5.4.0-dev. I can reproduce it in 5.3.0.

But a crash is a crash is a crash.

I have a suspicion this bug actually has nothing to do with dropped items per se, and something deeper is going on here. The error message is pretty weird, it is triggered by an `assert` that triggers if `moveresult` of `on_step` is `nil` for some reason.

This means that the `on_step` function of the entity is called without any `moveresult`.

There are two possibilities:
1. The `assert` is wrong. `moveresult` can very well be `nil`, and that's OK
2. Is is wrong that `on_step` was called without `moveresult` 

I have looked up the documentation and learned that the `moveresult` in `on_step` indeed might or might not be there. It says: “moveresult: Table with collision info (only available if physical=true)”. So I altered the code a bit to check what happens when I also check for `self.physical==true` before applying the assert. The assert still fires.

This means that `on_step` was indeed called with a `nil` `moveresult` although `self.physical` was set to `true`. I think this should never happen, according to the documentation `moveresult` MUST be set if `self.physical` is true.

Now the question is: How did it happen that `on_step` was called with a `nil` `moveresult`?

So the *real* bug, it seems to me, is that the API does not actually guarantee that `moveresult` is non-`nil` if `self.physical` is `true`, contrary to what `lua_api.txt` says. If this is the real bug, it actually has nothing to do with item entities. The bug only was discovered because of the assert.

**Comment by Wuzzy2** [comment]
There is also a third possibility: The object attacher tool from DevTest is broken. I have looked at this, but could not find anything suspicious. All this function really does to the object is call `set_attach`, and this call looks valid. So I don't think that's the reason.

**Comment by sfan5** [comment]
When an entity is attached to a player, collision detection will no longer run and `moveresult` will be nil.

**Comment by Wuzzy2** [comment]
So it turns out there are various valid situations in which `moveresult` can be legally `nil`. Wouldn't it make more sense to remove the `assert` entirely and instead just `return`? There's also a similar `assert` in falling nodes btw.
If the point of this `assert` is to test if `moveresult` is always "valid", I think it would make more sense to move this test to DevTest.

Oh, and this also implies the documentation is wrong/incomplete then. Because it does not say that `moveresult` can be `nil` if the entity is attached.

**Comment by rubenwardy** [comment]
the assert could also check for attachment with `get_attach`

**Comment by sfan5** [comment]
>  Wouldn't it make more sense to remove the assert entirely and instead just return?

No, because then it would no longer catch the cases where moveresult is nil because of broken games/mods.

**Comment by Wuzzy2** [comment]
But how could moveresult become nil because of broken games/mods? Isn't `on_step` called by Minetest rather than by mods? Can you give me an example of mod code that could break this?

> the assert could also check for attachment with get_attach

Yeah, but you would also have to check `physical`. The assert just becomes bigger and bigger and is run every step. Not sure if it's a good idea.

**Comment by rubenwardy** [comment]
> But how could moveresult become nil because of broken games/mods? Isn't on_step called by Minetest rather than by mods? Can you give me an example of mod code that could break this?

See https://github.com/minetest/minetest_game/commit/bd64188ee566bc31c1cd50d63f0da3328d42ab51
